cdc/controller/controller.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/orchestrator"
+	"github.com/pingcap/tiflow/pkg/upstream"
+	"github.com/pingcap/tiflow/pkg/version"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+)
+
+// versionInconsistentLogRate represents the rate of log output when there are
+// captures with versions different from that of the owner
+const versionInconsistentLogRate = 1
+
+// Controller is a manager to schedule changefeeds
+type Controller interface {
+	orchestrator.Reactor
+	AsyncStop()
+}
+
+type controllerImpl struct {
+	changefeeds     map[model.ChangeFeedID]*orchestrator.ChangefeedReactorState
+	upstreamManager *upstream.Manager
+
+	// logLimiter controls cluster version check log output rate
+	logLimiter   *rate.Limiter
+	lastTickTime time.Time
+	// bootstrapped specifies whether the owner has been initialized.
+	// This will only be done when the owner starts the first Tick.
+	// NOTICE: Do not use it in a method other than tick unexpectedly,
+	//         as it is not a thread-safe value.
+	bootstrapped bool
+
+	closed int32
+
+	cfg *config.SchedulerConfig
```

Comment:
Perhaps cfg can be removed from controller?

Reply:
removed